File: 2022/03/26/detectCapital.ts
/**
 * https://leetcode.com/problems/detect-capital/
 * @author Andrew Jarombek
 * @since 3/26/2022
 */

function detectCapitalUse(word: string): boolean {
    let words = word.split(" ");

    for (let word of words) {
        if (word.length > 1) {
            let first = word.charAt(0);
            let second = word.charAt(1);

            if (first.toUpperCase() == first && second.toUpperCase() == second) {
                for (let i = 2; i < word.length; i++) {
                    let c = word.charAt(i);
                    if (c.toUpperCase() != c) {
                        return false;
                    }
                }
            } else {
                for (let i = 1; i < word.length; i++) {
                    let c = word.charAt(i);
                    if (c.toUpperCase() == c) {
                        return false;
                    }
                }
            }
        }
    }

    return true;
}

File: 2023/07/10/go/detect_capital.go
/**
 * https://leetcode.com/problems/detect-capital/
 * Author: Andrew Jarombek
 * Date: 7/10/2023
 */

package _go

import (
	"unicode"
)

func detectCapitalUse(word string) bool {
	if len(word) < 2 {
		return true
	}

	first := rune(word[0])
	second := rune(word[1])

	if unicode.IsUpper(first) && unicode.IsUpper(second) {
		for i := 2; i < len(word); i++ {
			if !unicode.IsUpper(rune(word[i])) {
				return false
			}
		}
	} else {
		for i := 1; i < len(word); i++ {
			if unicode.IsUpper(rune(word[i])) {
				return false
			}
		}
	}

	return true
}
